{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up pipeline with basic imports, data directories, and file basenames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from benbiohelpers.DataPipelineManagement.GenomeManager import getGenomeFastaFilePath\n",
    "from bpde_chromatin_analysis.helper_scripts.BPDE_DataDir import getDataDir\n",
    "\n",
    "hg19FastaFilePath = getGenomeFastaFilePath(\"hg19\")\n",
    "\n",
    "adarDamageSeqDataDirectory = os.path.join(getDataDir(), \"Adar_damage-seq\")\n",
    "adarDamageSeqCellularBasename = \"A549_BPDE_cell_2h\"\n",
    "adarDamageSeqNakedBasename = \"A549_BPDE_nDNA_2h\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Format the reads\n",
    "Format the reads by combining paired-end reads, removing lone \"2nd pair\" reads, removing duplicates, and combining replicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bpde_chromatin_analysis.FormatAlignedDamageReads import formatAlignedDamageReads\n",
    "\n",
    "adarDamageSeqCellularReadsFilePaths = list()\n",
    "adarDamageSeqNakedReadsFilePaths = list()\n",
    "\n",
    "for rep in (\"_rep1\",\"_rep2\",\"_rep3\",\"_rep4\"):\n",
    "    adarDamageSeqCellularReadsFilePaths.append(os.path.join(adarDamageSeqDataDirectory, adarDamageSeqCellularBasename, adarDamageSeqCellularBasename+rep+\".bed\"))\n",
    "    adarDamageSeqNakedReadsFilePaths.append(os.path.join(adarDamageSeqDataDirectory, adarDamageSeqNakedBasename, adarDamageSeqNakedBasename+rep+\".bed\"))\n",
    "\n",
    "formattedAdarDamageSeqCellularReadsFilePaths = formatAlignedDamageReads(adarDamageSeqCellularReadsFilePaths)\n",
    "formattedAdarDamageSeqNakedReadsFilePaths = formatAlignedDamageReads(adarDamageSeqNakedReadsFilePaths)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert the BPDE damage-seq reads to single-nucleotide lesion positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bpde_chromatin_analysis.GetBDPE_DamagePositionsFromAlignedReads import getBPDE_DamagePositionsFromAlignedReads\n",
    "\n",
    "adarDamageSeqCellularDamagePosFilePaths = getBPDE_DamagePositionsFromAlignedReads(formattedAdarDamageSeqCellularReadsFilePaths, hg19FastaFilePath)\n",
    "adarDamageSeqNakedDamagePosFilePaths = getBPDE_DamagePositionsFromAlignedReads(formattedAdarDamageSeqNakedReadsFilePaths, hg19FastaFilePath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split damage positions into guanine and non-guanine positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bpde_chromatin_analysis.FilterOnSpecificNucleotides import filterOnSpecificNucleotides\n",
    "\n",
    "adarDamageSeqCellularDamagePosFilePaths = filterOnSpecificNucleotides(adarDamageSeqCellularDamagePosFilePaths, 'A')\n",
    "adarDamageSeqNakedDamagePosFilePaths = filterOnSpecificNucleotides(adarDamageSeqNakedDamagePosFilePaths, 'A')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Move results manually to \"A549_BPDE_cell_2h_A\" and \"A549_BPDE_nDNA_2h_A\" directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adarDamageSeqCellularBasename += \"_A\"\n",
    "adarDamageSeqNakedBasename += \"_A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse the damage position data for mutperiod input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mutperiodpy.input_parsing.ParseStandardBed import parseStandardBed\n",
    "\n",
    "adarDamageSeqCellularMutperiodInputFilePath = parseStandardBed(adarDamageSeqCellularDamagePosFilePaths, hg19FastaFilePath)[0]\n",
    "adarDamageSeqNakedMutperiodInputFilePath = parseStandardBed(adarDamageSeqNakedDamagePosFilePaths, hg19FastaFilePath)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieve processed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from benbiohelpers.FileSystemHandling.DirectoryHandling import getFilesInDirectory\n",
    "from mutperiodpy.helper_scripts.UsefulFileSystemFunctions import getExternalDataDirectory as getMutperiodExternalDataDirectory, DataTypeStr\n",
    "\n",
    "mutperiodHg19Directory = getMutperiodExternalDataDirectory()\n",
    "\n",
    "adarDamageSeqCellularMutperiodInputFilePaths = getFilesInDirectory(os.path.join(adarDamageSeqDataDirectory, adarDamageSeqCellularBasename), DataTypeStr.mutations + \".bed\")\n",
    "adarDamageSeqNakedMutperiodInputFilePath = getFilesInDirectory(os.path.join(adarDamageSeqDataDirectory, adarDamageSeqNakedBasename), DataTypeStr.mutations + \".bed\", searchRecursively = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nucleosome Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TFBS Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mutperiodpy.RunAnalysisSuite import runAnalysisSuite, generateCustomBackground\n",
    "\n",
    "TFBS_FilePaths = list()\n",
    "TFBS_FilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_CTCF_known\", \"hg19_CTCF_known.bed\"))\n",
    "# TFBS_FilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_ETS_and_GABPA_known\", \"hg19_ETS_and_GABPA_known.bed\"))\n",
    "TFBS_FilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_SP1_known\", \"hg19_SP1_known.bed\"))\n",
    "\n",
    "TFBS_Names = [os.path.basename(TFBS_FilePath).rsplit('.', 1)[0] for TFBS_FilePath in TFBS_FilePaths]\n",
    "\n",
    "generateCustomBackground(os.path.dirname(adarDamageSeqNakedMutperiodInputFilePath), TFBS_Names,\n",
    "                         useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True, useNucStrand = True)\n",
    "\n",
    "runAnalysisSuite(adarDamageSeqCellularMutperiodInputFilePaths, TFBS_Names,\n",
    "                 normalizationMethod = \"Custom Background\", customBackgroundDir = os.path.dirname(adarDamageSeqNakedMutperiodInputFilePath),\n",
    "                 useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True, useNucStrand = True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
