{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up pipeline with basic imports, data directories, and file basenames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from benbiohelpers.DataPipelineManagement.GenomeManager import getGenomeFastaFilePath\n",
    "from bpde_chromatin_analysis.helper_scripts.BPDE_DataDir import getDataDir\n",
    "from mutperiodpy.helper_scripts.UsefulFileSystemFunctions import getExternalDataDirectory as getMutperiodExternalDataDirectory\n",
    "\n",
    "hg19FastaFilePath = getGenomeFastaFilePath(\"hg19\")\n",
    "mutperiodHg19Directory = getMutperiodExternalDataDirectory()\n",
    "\n",
    "liDataDirectory = os.path.join(getDataDir(), \"Li_tXR-seq\")\n",
    "liBaseDataFilePath = os.path.join(liDataDirectory, \"Li_tXR-seq.bed\")\n",
    "\n",
    "adarDamageSeqDataDir = os.path.join(getDataDir(), \"Adar_damage-seq\", \"A549_BPDE_cell_2h\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse the tXR-seq data for mutperiod input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found parsed data at /home/ben/bioinformatics/bpde_chromatin_analysis/BPDE_data/Li_tXR-seq/Li_tXR-seq_trinuc_context_mutations.bed. Continuing.\n"
     ]
    }
   ],
   "source": [
    "from mutperiodpy.input_parsing.ParseXRSeq import parseXRSeq, PresetCallParams\n",
    "from mutperiodpy.helper_scripts.UsefulFileSystemFunctions import DataTypeStr\n",
    "from benbiohelpers.FileSystemHandling.DirectoryHandling import getFilesInDirectory\n",
    "\n",
    "if getFilesInDirectory(liDataDirectory, DataTypeStr.mutations + \".bed\", searchRecursively = False) is None:\n",
    "    print(\"Parsed data not found. Generating...\")\n",
    "    liMutperiodInputFilePaths = parseXRSeq([liBaseDataFilePath], PresetCallParams.BPDE, hg19FastaFilePath)\n",
    "else:\n",
    "    liMutperiodInputFilePaths = [getFilesInDirectory(liDataDirectory, DataTypeStr.mutations + \".bed\", searchRecursively = False)]\n",
    "    print(f\"Found parsed data at {liMutperiodInputFilePaths[0]}. Continuing.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the mutperiod pipeline to determine translational and rotational periodicity of the data.\n",
    "NOTE: This analysis requires that the damage data has already been parsed for mutperiod using the Adar_damage-seq_nucleosome_periodicity_data_generation notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counting mutations at each dyad position...\n",
      "\n",
      "Working with Li_tXR-seq_trinuc_context_mutations.bed\n",
      "Counting with nucleosome map: hg19_hybrid_nuc_map\n",
      "Counting mutations at each nucleosome position in a 73 bp radius + 0 bp linker DNA.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting mutations at each nucleosome position in a 1000 bp radius.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting with nucleosome map: hg19_LCL_MNase_nuc_map\n",
      "Counting mutations at each nucleosome position in a 73 bp radius + 0 bp linker DNA.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting mutations at each nucleosome position in a 1000 bp radius.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "\n",
      "Normalizing counts using custom background data...\n",
      "\n",
      "Working with Li_tXR-seq_hybrid_nuc_map_raw_nucleosome_mutation_counts.tsv and A549_BPDE_cell_2h_hybrid_nuc_map_raw_nucleosome_mutation_counts.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_hybrid_nuc_map_nuc-group_raw_nucleosome_mutation_counts.tsv and A549_BPDE_cell_2h_hybrid_nuc_map_nuc-group_raw_nucleosome_mutation_counts.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_LCL_MNase_map_raw_nucleosome_mutation_counts.tsv and A549_BPDE_cell_2h_LCL_MNase_map_raw_nucleosome_mutation_counts.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_LCL_MNase_map_nuc-group_raw_nucleosome_mutation_counts.tsv and A549_BPDE_cell_2h_LCL_MNase_map_nuc-group_raw_nucleosome_mutation_counts.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Expanding file context where necessary...\n",
      "\n",
      "\n",
      "Counting mutations at each dyad position...\n",
      "\n",
      "Working with Li_tXR-seq_trinuc_context_mutations.bed\n",
      "Counting with nucleosome map: hg19_hybrid_nuc_map\n",
      "Counting mutations at each nucleosome position in a 73 bp radius + 0 bp linker DNA.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting mutations at each nucleosome position in a 1000 bp radius.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting with nucleosome map: hg19_LCL_MNase_nuc_map\n",
      "Counting mutations at each nucleosome position in a 73 bp radius + 0 bp linker DNA.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "Counting mutations at each nucleosome position in a 1000 bp radius.\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in chr1\n",
      "Counting in chr10\n",
      "Counting in chr11\n",
      "Counting in chr12\n",
      "Counting in chr13\n",
      "Counting in chr14\n",
      "Counting in chr15\n",
      "Counting in chr16\n",
      "Counting in chr17\n",
      "Counting in chr18\n",
      "Counting in chr19\n",
      "Counting in chr2\n",
      "Counting in chr20\n",
      "Counting in chr21\n",
      "Counting in chr22\n",
      "Counting in chr3\n",
      "Counting in chr4\n",
      "Counting in chr5\n",
      "Counting in chr6\n",
      "Counting in chr7\n",
      "Counting in chr8\n",
      "Counting in chr9\n",
      "Counting in chrX\n",
      "\n",
      "Generating genome-wide mutation background...\n",
      "\n",
      "Working in: Li_tXR-seq_trinuc_context_mutations.bed\n",
      "Generating mutation context frequency file...\n",
      "Generating mutation background...\n",
      "\n",
      "Generating nucleosome mutation background...\n",
      "\n",
      "Working with Li_tXR-seq_trinuc_mutation_background.tsv\n",
      "Counting with nucleosome map: hg19_hybrid_nuc_map\n",
      "Given mutation background is in trinuc context.\n",
      "Counting with nucleosome map: hg19_LCL_MNase_nuc_map\n",
      "Given mutation background is in trinuc context.\n",
      "\n",
      "Normalizing counts with nucleosome background data...\n",
      "\n",
      "Working with Li_tXR-seq_hybrid_nuc_map_raw_nucleosome_mutation_counts.tsv and Li_tXR-seq_hybrid_nuc_map_trinuc_nucleosome_mutation_background.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_hybrid_nuc_map_nuc-group_raw_nucleosome_mutation_counts.tsv and Li_tXR-seq_hybrid_nuc_map_trinuc_nuc-group_nucleosome_mutation_background.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_LCL_MNase_map_raw_nucleosome_mutation_counts.tsv and Li_tXR-seq_LCL_MNase_map_trinuc_nucleosome_mutation_background.tsv\n",
      "Calling R script to generate normalized counts...\n",
      "\n",
      "Working with Li_tXR-seq_LCL_MNase_map_nuc-group_raw_nucleosome_mutation_counts.tsv and Li_tXR-seq_LCL_MNase_map_trinuc_nuc-group_nucleosome_mutation_background.tsv\n",
      "Calling R script to generate normalized counts...\n"
     ]
    }
   ],
   "source": [
    "from mutperiodpy.RunAnalysisSuite import runAnalysisSuite\n",
    "nucleosomeMapFilePaths = list()\n",
    "nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_hybrid_nuc_map\", \"hg19_hybrid_nuc_map.bed\"))\n",
    "nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_LCL_MNase_nuc_map\", \"hg19_LCL_MNase_nuc_map.bed\"))\n",
    "# nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_LCL_MNase_nuc_map_euchromatin\",\n",
    "#                                            \"hg19_LCL_MNase_nuc_map_euchromatin.bed\"))\n",
    "# nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_LCL_MNase_nuc_map_heterochromatin\",\n",
    "#                                            \"hg19_LCL_MNase_nuc_map_heterochromatin.bed\"))\n",
    "# nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_LCL_MNase_nuc_map_stringent_euchromatin\",\n",
    "#                                            \"hg19_LCL_MNase_nuc_map_stringent_euchromatin.bed\"))\n",
    "# nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_NHF1_MNase_nuc_map\", \"hg19_NHF1_MNase_nuc_map.bed\"))\n",
    "\n",
    "nucleosomeMapNames = [os.path.basename(nucleosomeMapFilePath).rsplit('.', 1)[0] for nucleosomeMapFilePath in nucleosomeMapFilePaths]\n",
    "\n",
    "# runAnalysisSuite(liMutperiodInputFilePaths, nucleosomeMapNames,\n",
    "#                  normalizationMethod = \"No Normalization\", customBackgroundDir = None,\n",
    "#                  useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True)\n",
    "# runAnalysisSuite(liMutperiodInputFilePaths, nucleosomeMapNames,\n",
    "#                  normalizationMethod = \"Singlenuc/Dinuc\", customBackgroundDir = None,\n",
    "#                  useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True)\n",
    "runAnalysisSuite(liMutperiodInputFilePaths, nucleosomeMapNames,\n",
    "                 normalizationMethod = \"Custom Background\", customBackgroundDir = adarDamageSeqDataDir,\n",
    "                 useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True)\n",
    "runAnalysisSuite(liMutperiodInputFilePaths, nucleosomeMapNames,\n",
    "                 normalizationMethod = \"Trinuc/Quadrunuc\", customBackgroundDir = None,\n",
    "                 useSingleNucRadius = True, includeLinker = False, useNucGroupRadius = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating inputs to run analysis without grouped comparison...\n",
      "Calling R script...\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_nuc-group_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_nuc-group_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "Results can be found at /home/ben/bioinformatics/bpde_chromatin_analysis/BPDE_data/Li_tXR-seq/Li_tXR-seq_raw_periodicity_data.tsv\n",
      "Generating inputs to run analysis without grouped comparison...\n",
      "Calling R script...\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_nuc-group_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_nuc-group_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_raw\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "Results can be found at /home/ben/bioinformatics/bpde_chromatin_analysis/BPDE_data/Li_tXR-seq/Li_tXR-seq_raw_aligned_periodicity_data.tsv\n",
      "Generating inputs to run analysis without grouped comparison...\n",
      "Calling R script...\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_custom_context_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_custom_context_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_singlenuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_singlenuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_trinuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_trinuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_custom_context_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_custom_context_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_singlenuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_singlenuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_trinuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_trinuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "Results can be found at /home/ben/bioinformatics/bpde_chromatin_analysis/BPDE_data/Li_tXR-seq/Li_tXR-seq_normalized_periodicity_data.tsv\n",
      "Generating inputs to run analysis without grouped comparison...\n",
      "Calling R script...\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_custom_context_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_custom_context_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_singlenuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_singlenuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_trinuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_LCL_MNase_map_trinuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_custom_context_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_custom_context_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_singlenuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_singlenuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_trinuc_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "[1] \"Working with Li_tXR-seq_hybrid_nuc_map_trinuc_nuc-group_normalized\"\n",
      "[1] \"Running periodicity analysis...\"\n",
      "Results can be found at /home/ben/bioinformatics/bpde_chromatin_analysis/BPDE_data/Li_tXR-seq/Li_tXR-seq_normalized_aligned_periodicity_data.tsv\n"
     ]
    }
   ],
   "source": [
    "from mutperiodpy.RunNucleosomeMutationAnalysis import runNucleosomeMutationAnalysis\n",
    "\n",
    "# rawNucleosomeCountsFilePaths = list()\n",
    "# for nucleosomeMapName in nucleosomeMapNames:\n",
    "#     rawNucleosomeCountsFilePaths += getFilesInDirectory(os.path.join(liDataDirectory, nucleosomeMapName), DataTypeStr.rawNucCounts + \".tsv\")\n",
    "# runNucleosomeMutationAnalysis(rawNucleosomeCountsFilePaths,\n",
    "#                               outputFilePath = os.path.join(liDataDirectory, \"Li_tXR-seq_raw_periodicity_data.tsv\"),\n",
    "#                               overridePeakPeriodicityWithExpected = False, alignStrands = False)\n",
    "# runNucleosomeMutationAnalysis(rawNucleosomeCountsFilePaths,\n",
    "#                               outputFilePath = os.path.join(liDataDirectory, \"Li_tXR-seq_raw_aligned_periodicity_data.tsv\"),\n",
    "#                               overridePeakPeriodicityWithExpected = False, alignStrands = True)\n",
    "\n",
    "normalizedNucleosomeCountsFilePaths = list()\n",
    "for nucleosomeMapName in nucleosomeMapNames:\n",
    "    normalizedNucleosomeCountsFilePaths += getFilesInDirectory(os.path.join(liDataDirectory, nucleosomeMapName), DataTypeStr.normNucCounts + \".tsv\")\n",
    "runNucleosomeMutationAnalysis(normalizedNucleosomeCountsFilePaths,\n",
    "                              outputFilePath = os.path.join(liDataDirectory, \"Li_tXR-seq_normalized_periodicity_data.tsv\"),\n",
    "                              overridePeakPeriodicityWithExpected = False, alignStrands = False)\n",
    "runNucleosomeMutationAnalysis(normalizedNucleosomeCountsFilePaths,\n",
    "                              outputFilePath = os.path.join(liDataDirectory, \"Li_tXR-seq_normalized_aligned_periodicity_data.tsv\"),\n",
    "                              overridePeakPeriodicityWithExpected = False, alignStrands = True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
