{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "698956af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from benbiohelpers.DataPipelineManagement.GenomeManager import getGenomeFastaFilePath\n",
    "from bpde_chromatin_analysis.helper_scripts.BPDE_DataDir import getDataDir\n",
    "from mutperiodpy.helper_scripts.UsefulFileSystemFunctions import getExternalDataDirectory as getMutperiodExternalDataDirectory\n",
    "\n",
    "hg19FastaFilePath = getGenomeFastaFilePath(\"hg19\")\n",
    "mutperiodHg19Directory = os.path.join(getMutperiodExternalDataDirectory(), \"hg19\")\n",
    "\n",
    "smmSeqDataDirectory = os.path.join(getDataDir(), \"Adar_smm-seq\")\n",
    "smmSeqBaseDataFilePath = os.path.join(smmSeqDataDirectory, \"smm-seq_BPDE_all_reps.bed\")\n",
    "smmSeqTrinucDataFilePath = os.path.join(smmSeqDataDirectory, \"smm-seq_BPDE_all_reps_trinuc_context.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a6af997",
   "metadata": {},
   "source": [
    "### Generate trinucleotide mutation counts inside and outside nucleosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe01984d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from benbiohelpers.CountThisInThat.Counter import ThisInThatCounter\n",
    "from benbiohelpers.CountThisInThat.CounterOutputDataHandler import ENCOMPASSED_DATA, OutputDataWriter, CounterOutputDataHandler\n",
    "from benbiohelpers.CountThisInThat.InputDataStructures import EncompassingDataDefaultStrand\n",
    "\n",
    "from benbiohelpers.FileSystemHandling.DirectoryHandling import checkDirs\n",
    "\n",
    "def getCountDerivatives(outputDataWriter: OutputDataWriter, getHeaders):\n",
    "    if getHeaders: return [\"Ratio\"]\n",
    "    else:\n",
    "        if outputDataWriter.totalCounts == 0: return 0\n",
    "        else: return [str(outputDataWriter.outputDataStructure[outputDataWriter.previousKeys[0]][outputDataWriter.previousKeys[1]][None]/outputDataWriter.totalCounts)]\n",
    "\n",
    "class MutationsInNucleosomes(ThisInThatCounter):\n",
    "\n",
    "    def constructEncompassingFeature(self, line) -> EncompassingDataDefaultStrand:\n",
    "        return EncompassingDataDefaultStrand(line, self.acceptableChromosomes)\n",
    "\n",
    "    def initOutputDataHandler(self):\n",
    "        \"\"\"\n",
    "        Use this function to create the instance of the CounterOutputDataHandler object.\n",
    "        Default behavior creates the output data handler and passes in the counter's \"writeIncrementally\" value.\n",
    "        \"\"\"\n",
    "        self.outputDataHandler = CounterOutputDataHandler(self.writeIncrementally)\n",
    "\n",
    "    def setupOutputDataStratifiers(self):\n",
    "        self.outputDataHandler.addSimpleEncompassedColStratifier(outputName=\"Mutated_From\", colIndex=3)\n",
    "        self.outputDataHandler.addSimpleEncompassedColStratifier(outputName=\"Mutated_To\", colIndex=4)\n",
    "        self.outputDataHandler.addPlaceholderStratifier(outputName=\"Counts\")\n",
    "\n",
    "    def setupOutputDataWriter(self):\n",
    "        self.outputDataHandler.createOutputDataWriter(self.outputFilePath, getCountDerivatives = getCountDerivatives, omitZeroRows = True)\n",
    "\n",
    "class MutationsOutsideNucleosomes(ThisInThatCounter):\n",
    "\n",
    "    def constructEncompassingFeature(self, line) -> EncompassingDataDefaultStrand:\n",
    "        return EncompassingDataDefaultStrand(line, self.acceptableChromosomes)\n",
    "\n",
    "    def initOutputDataHandler(self):\n",
    "        \"\"\"\n",
    "        Use this function to create the instance of the CounterOutputDataHandler object.\n",
    "        Default behavior creates the output data handler and passes in the counter's \"writeIncrementally\" value.\n",
    "        \"\"\"\n",
    "        self.outputDataHandler = CounterOutputDataHandler(self.writeIncrementally, trackAllEncompassed=True, countOnlyNonEncompassed=True)\n",
    "\n",
    "    def setupOutputDataStratifiers(self):\n",
    "        self.outputDataHandler.addSimpleEncompassedColStratifier(outputName=\"Mutated_From\", colIndex=3)\n",
    "        self.outputDataHandler.addSimpleEncompassedColStratifier(outputName=\"Mutated_To\", colIndex=4)\n",
    "        self.outputDataHandler.addPlaceholderStratifier(outputName=\"Counts\")\n",
    "\n",
    "    def setupOutputDataWriter(self):\n",
    "        self.outputDataHandler.createOutputDataWriter(self.outputFilePath, getCountDerivatives = getCountDerivatives, omitZeroRows = True)\n",
    "\n",
    "nucleosomeMapFilePaths = list()\n",
    "nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_hybrid_nuc_map\", \"hg19_hybrid_nuc_map.bed\"))\n",
    "nucleosomeMapFilePaths.append(os.path.join(mutperiodHg19Directory, \"hg19_LCL_MNase_nuc_map\", \"hg19_LCL_MNase_nuc_map.bed\"))\n",
    "\n",
    "mutationsInNucleosomesFilePaths = dict()\n",
    "mutationsOutsideNucleosomesFilePaths = dict()\n",
    "for nucleosomeMapFilePath in nucleosomeMapFilePaths:\n",
    "    nucleosomeMapName = os.path.basename(nucleosomeMapFilePath).rsplit('.', 1)[0]\n",
    "    checkDirs(os.path.join(smmSeqDataDirectory, nucleosomeMapName))\n",
    "    outputFilePath = os.path.join(smmSeqDataDirectory, nucleosomeMapName, \"mutations_in_\" + nucleosomeMapName + \".tsv\")\n",
    "    MutationsInNucleosomes(smmSeqBaseDataFilePath, nucleosomeMapFilePath, outputFilePath,\n",
    "                           encompassingFeatureExtraRadius=73).count()\n",
    "    outputFilePath = os.path.join(smmSeqDataDirectory, nucleosomeMapName, \"trinuc_mutations_in_\" + nucleosomeMapName + \".tsv\")\n",
    "    MutationsInNucleosomes(smmSeqTrinucDataFilePath, nucleosomeMapFilePath, outputFilePath,\n",
    "                           encompassingFeatureExtraRadius=73).count()\n",
    "    mutationsInNucleosomesFilePaths[nucleosomeMapFilePath] = outputFilePath\n",
    "\n",
    "    outputFilePath = os.path.join(smmSeqDataDirectory, nucleosomeMapName, \"mutations_outside_\" + nucleosomeMapName + \".tsv\")\n",
    "    MutationsOutsideNucleosomes(smmSeqBaseDataFilePath, nucleosomeMapFilePath, outputFilePath,\n",
    "                                encompassingFeatureExtraRadius=73).count()\n",
    "    outputFilePath = os.path.join(smmSeqDataDirectory, nucleosomeMapName, \"trinuc_mutations_outside_\" + nucleosomeMapName + \".tsv\")\n",
    "    MutationsOutsideNucleosomes(smmSeqTrinucDataFilePath, nucleosomeMapFilePath, outputFilePath,\n",
    "                                encompassingFeatureExtraRadius=73).count()\n",
    "    mutationsOutsideNucleosomesFilePaths[nucleosomeMapFilePath] = outputFilePath\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b93ef505",
   "metadata": {},
   "source": [
    "### Fill in missing trinucleotide counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f438ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "for nucleosomeMapFilePath in nucleosomeMapFilePaths:\n",
    "    mutationsInNucleosomes = pd.read_table(mutationsInNucleosomesFilePaths[nucleosomeMapFilePath])\n",
    "    mutationsOutsideNucleosomes = pd.read_table(mutationsOutsideNucleosomesFilePaths[nucleosomeMapFilePath])\n",
    "    missedContextsInNucleosomes = []\n",
    "    missedContextsOutsideNucleosomes = []\n",
    "    for fivePrimeNucleotide in ['A','C','T','G']:\n",
    "        for mutatedFrom in ['C','T']:\n",
    "            for threePrimeNucleotide in ['A','C','T','G']:\n",
    "                trinucleotide = fivePrimeNucleotide + mutatedFrom + threePrimeNucleotide\n",
    "                mutInNucTrinucRows = mutationsInNucleosomes[mutationsInNucleosomes[\"Mutated_From\"]==trinucleotide]\n",
    "                mutOutsideNucTrinucRows = mutationsOutsideNucleosomes[mutationsOutsideNucleosomes[\"Mutated_From\"]==trinucleotide]\n",
    "                for mutatedTo in ['A','C','T','G']:\n",
    "                    if mutatedTo == mutatedFrom: continue\n",
    "                    if mutatedTo not in mutInNucTrinucRows[\"Mutated_To\"].values:\n",
    "                        missedContextsInNucleosomes.append([trinucleotide, mutatedTo, 0, 0])\n",
    "                    if mutatedTo not in mutOutsideNucTrinucRows[\"Mutated_To\"].values:\n",
    "                        missedContextsOutsideNucleosomes.append([trinucleotide, mutatedTo, 0, 0])\n",
    "    missedContextsInNucleosomes = pd.DataFrame(missedContextsInNucleosomes, columns=mutationsInNucleosomes.columns)\n",
    "    missedContextsOutsideNucleosomes = pd.DataFrame(missedContextsOutsideNucleosomes, columns=mutationsOutsideNucleosomes.columns)\n",
    "    mutationsInNucleosomes = pd.concat([mutationsInNucleosomes, missedContextsInNucleosomes], ignore_index=True).sort_values(by=[\"Mutated_From\", \"Mutated_To\"])\n",
    "    mutationsOutsideNucleosomes = pd.concat([mutationsOutsideNucleosomes, missedContextsOutsideNucleosomes], ignore_index=True).sort_values(by=[\"Mutated_From\", \"Mutated_To\"])\n",
    "    mutationsInNucleosomes.to_csv(mutationsInNucleosomesFilePaths[nucleosomeMapFilePath], sep = '\\t', index = False)\n",
    "    mutationsOutsideNucleosomes.to_csv(mutationsOutsideNucleosomesFilePaths[nucleosomeMapFilePath], sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4c2a2dd",
   "metadata": {},
   "source": [
    "### Compute cosine similarity on mutation signatures inside and outside nucleosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb6acdb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy.linalg import norm\n",
    "for nucleosomeMapFilePath in nucleosomeMapFilePaths:\n",
    "    mutationsInNucleosomes = pd.read_table(mutationsInNucleosomesFilePaths[nucleosomeMapFilePath])\n",
    "    mutationsOutsideNucleosomes = pd.read_table(mutationsOutsideNucleosomesFilePaths[nucleosomeMapFilePath])\n",
    "    mergedMutations = mutationsInNucleosomes.merge(\n",
    "        mutationsOutsideNucleosomes, \"outer\", [\"Mutated_From\",\"Mutated_To\"], sort = True, suffixes=(\"_In_Nucleosomes\", \"_Outside_Nucleosomes\")\n",
    "    ).fillna(0)\n",
    "    print(os.path.basename(nucleosomeMapFilePath).rsplit('.', 1)[0],\":\",sep='')\n",
    "    ratioIn = np.array(mergedMutations[\"Ratio_In_Nucleosomes\"])\n",
    "    ratioOut = np.array(mergedMutations[\"Ratio_Outside_Nucleosomes\"])\n",
    "    cosineSimilarity = np.dot(ratioIn, ratioOut) / (norm(ratioIn) * norm(ratioOut))\n",
    "    print(\"Cosine similarity (mutations inside vs. outside nucleosomes):\",cosineSimilarity)\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
